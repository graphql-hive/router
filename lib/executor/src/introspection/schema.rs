use std::collections::{HashMap, HashSet};

use graphql_parser::{
    query::Type,
    schema::{Definition, TypeDefinition},
};
use hive_router_query_planner::consumer_schema::ConsumerSchema;

#[derive(Debug, Default)]
pub struct SchemaMetadata {
    pub possible_types: PossibleTypes,
    pub enum_values: HashMap<String, HashSet<String>>,
    pub type_fields: HashMap<String, HashMap<String, String>>,
    pub object_types: HashSet<String>,
    pub scalar_types: HashSet<String>,
}

impl SchemaMetadata {
    pub fn is_object_type(&self, name: &str) -> bool {
        self.object_types.contains(name)
    }

    pub fn is_scalar_type(&self, name: &str) -> bool {
        self.scalar_types.contains(name)
    }
}

#[derive(Debug, Default)]
pub struct PossibleTypes {
    map: HashMap<String, HashSet<String>>,
}

impl PossibleTypes {
    pub fn entity_satisfies_type_condition(&self, type_name: &str, type_condition: &str) -> bool {
        if type_name == type_condition {
            true
        } else if let Some(possible_types_of_type) = self.map.get(type_condition) {
            possible_types_of_type.contains(type_name)
        } else {
            false
        }
    }
    pub fn get_possible_types(&self, type_name: &str) -> HashSet<String> {
        let mut possible_types = self.map.get(type_name).cloned().unwrap_or_default();
        possible_types.insert(type_name.to_string());
        possible_types
    }
}

pub trait SchemaWithMetadata {
    fn schema_metadata(&self) -> SchemaMetadata;
}

impl SchemaWithMetadata for ConsumerSchema {
    fn schema_metadata(&self) -> SchemaMetadata {
        let mut first_possible_types: HashMap<String, Vec<String>> = HashMap::new();
        let mut type_fields: HashMap<String, HashMap<String, String>> = HashMap::new();
        let mut enum_values: HashMap<String, HashSet<String>> = HashMap::new();
        let mut scalar_types: HashSet<String> = HashSet::from_iter(vec![
            "Boolean".to_string(),
            "Float".to_string(),
            "Int".to_string(),
            "ID".to_string(),
            "String".to_string(),
        ]);
        let mut object_types: HashSet<String> = HashSet::new();

        for definition in &self.document.definitions {
            match definition {
                Definition::TypeDefinition(TypeDefinition::Enum(enum_type)) => {
                    let name = enum_type.name.to_string();
                    let mut values = HashSet::new();
                    for enum_value in &enum_type.values {
                        values.insert(enum_value.name.to_string());
                    }
                    enum_values.insert(name, values);
                }
                Definition::TypeDefinition(TypeDefinition::Object(object_type)) => {
                    let name = object_type.name.to_string();
                    object_types.insert(name.clone());
                    let fields = type_fields.entry(name).or_default();
                    for field in &object_type.fields {
                        let field_type_name = field.field_type.type_name();
                        fields.insert(field.name.to_string(), field_type_name);
                    }

                    for interface in &object_type.implements_interfaces {
                        let interface_name = interface.to_string();
                        let possible_types_entry =
                            first_possible_types.entry(interface_name).or_default();
                        possible_types_entry.push(object_type.name.to_string());
                    }
                }
                Definition::TypeDefinition(TypeDefinition::Interface(interface_type)) => {
                    let name = interface_type.name.to_string();
                    let mut fields = HashMap::new();
                    for field in &interface_type.fields {
                        let field_type_name = field.field_type.type_name();
                        fields.insert(field.name.to_string(), field_type_name);
                    }
                    type_fields.insert(name, fields);
                    for interface_name in &interface_type.implements_interfaces {
                        let interface_name = interface_name.to_string();
                        let possible_types_entry =
                            first_possible_types.entry(interface_name).or_default();
                        possible_types_entry.push(interface_type.name.to_string());
                    }
                }
                Definition::TypeDefinition(TypeDefinition::Union(union_type)) => {
                    let name = union_type.name.to_string();
                    let mut types = vec![];
                    for member in &union_type.types {
                        types.push(member.to_string());
                    }
                    first_possible_types.insert(name, types);
                }
                Definition::TypeDefinition(TypeDefinition::Scalar(scalar_type)) => {
                    scalar_types.insert(scalar_type.name.to_string());
                }
                _ => {}
            }
        }

        let mut final_possible_types: HashMap<String, HashSet<String>> = HashMap::new();
        // Re-iterate over the possible_types
        for (definition_name_of_x, first_possible_types_of_x) in &first_possible_types {
            let mut possible_types_of_x: HashSet<String> = HashSet::new();
            for definition_name_of_y in first_possible_types_of_x {
                possible_types_of_x.insert(definition_name_of_y.to_string());
                let possible_types_of_y = first_possible_types.get(definition_name_of_y);
                if let Some(possible_types_of_y) = possible_types_of_y {
                    for definition_name_of_z in possible_types_of_y {
                        possible_types_of_x.insert(definition_name_of_z.to_string());
                    }
                }
            }
            final_possible_types.insert(definition_name_of_x.to_string(), possible_types_of_x);
        }

        SchemaMetadata {
            possible_types: PossibleTypes {
                map: final_possible_types,
            },
            enum_values,
            type_fields,
            object_types,
            scalar_types,
        }
    }
}

trait TypeName {
    fn type_name(&self) -> String;
}

impl TypeName for Type<'_, String> {
    fn type_name(&self) -> String {
        match self {
            graphql_parser::schema::Type::NamedType(named_type) => named_type.to_string(),
            graphql_parser::schema::Type::NonNullType(non_null_type) => non_null_type.type_name(),
            graphql_parser::schema::Type::ListType(list_type) => list_type.type_name(),
        }
    }
}
