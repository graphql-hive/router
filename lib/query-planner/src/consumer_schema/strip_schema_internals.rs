use graphql_parser::query::Text;
use graphql_parser::schema::*;

use crate::{
    federation_spec::{
        definitions::{
            CorePurposesEnum, JoinFieldSetScalar, JoinGraphEnum, LinkImportScalar, LinkPurposeEnum,
        },
        directives::{
            CoreDirective, InaccessibleDirective, JoinEnumValueDirective, JoinFieldDirective,
            JoinGraphDirective, JoinImplementsDirective, JoinTypeDirective,
            JoinUnionMemberDirective, LinkDirective, TagDirective,
        },
        join_owner::JoinOwnerDirective,
    },
    utils::schema_transformer::{SchemaTransformer, TransformedValue},
};

// directive @inaccessible on FIELD_DEFINITION | OBJECT | INTERFACE | UNION | ENUM | ENUM_VALUE | SCALAR | INPUT_OBJECT | INPUT_FIELD_DEFINITION | ARGUMENT_DEFINITION
pub(crate) struct StripSchemaInternals;

static DIRECTIVES_TO_STRIP: [&str; 11] = [
    JoinTypeDirective::NAME,
    JoinEnumValueDirective::NAME,
    JoinFieldDirective::NAME,
    JoinImplementsDirective::NAME,
    JoinUnionMemberDirective::NAME,
    JoinGraphDirective::NAME,
    JoinOwnerDirective::NAME,
    LinkDirective::NAME,
    TagDirective::NAME,
    InaccessibleDirective::NAME,
    CoreDirective::NAME,
];

static DEFINITIONS_TO_STRIP: [&str; 5] = [
    LinkPurposeEnum::NAME,
    LinkImportScalar::NAME,
    JoinGraphEnum::NAME,
    JoinFieldSetScalar::NAME,
    CorePurposesEnum::NAME,
];

impl StripSchemaInternals {
    pub fn strip_schema_internals(schema: &Document<'static, String>) -> Document<'static, String> {
        let mut transformer = StripSchemaInternals {};
        let result = transformer
            .transform_document(schema)
            .replace_or_else(|| schema.clone());

        result
    }

    pub(crate) fn filter_directives<'a, T: Text<'a> + Clone>(
        directives: &Vec<Directive<'a, T>>,
    ) -> Vec<Directive<'a, T>> {
        directives
            .iter()
            .filter(|d| !DIRECTIVES_TO_STRIP.contains(&d.name.as_ref()))
            .cloned()
            .collect()
    }
}

impl<'a, T: Text<'a> + Clone> SchemaTransformer<'a, T> for StripSchemaInternals {
    fn transform_directives(
        &mut self,
        directives: &Vec<Directive<'a, T>>,
    ) -> TransformedValue<Vec<Directive<'a, T>>> {
        let new_directives = StripSchemaInternals::filter_directives(directives);

        TransformedValue::Replace(new_directives)
    }

    fn transform_document(
        &mut self,
        document: &Document<'a, T>,
    ) -> TransformedValue<Document<'a, T>> {
        let new_doc = Document {
            definitions: document
                .definitions
                .iter()
                .filter(|def| match def {
                    Definition::DirectiveDefinition(directive) => {
                        !DIRECTIVES_TO_STRIP.contains(&directive.name.as_ref())
                    }
                    Definition::TypeDefinition(TypeDefinition::Object(obj)) => {
                        !DEFINITIONS_TO_STRIP.contains(&obj.name.as_ref())
                    }
                    Definition::TypeDefinition(TypeDefinition::Interface(interface)) => {
                        !DEFINITIONS_TO_STRIP.contains(&interface.name.as_ref())
                    }
                    Definition::TypeDefinition(TypeDefinition::Union(union)) => {
                        !DEFINITIONS_TO_STRIP.contains(&union.name.as_ref())
                    }
                    Definition::TypeDefinition(TypeDefinition::Enum(enm)) => {
                        !DEFINITIONS_TO_STRIP.contains(&enm.name.as_ref())
                    }
                    Definition::TypeDefinition(TypeDefinition::Scalar(scalar)) => {
                        !DEFINITIONS_TO_STRIP.contains(&scalar.name.as_ref())
                    }
                    Definition::TypeDefinition(TypeDefinition::InputObject(input)) => {
                        !DEFINITIONS_TO_STRIP.contains(&input.name.as_ref())
                    }
                    Definition::TypeExtension(_) => todo!("TypeExtension not implemented"),
                    Definition::SchemaDefinition(_) => true,
                })
                .cloned()
                .collect(),
        };

        self.default_transform_document(&new_doc)
    }
}

#[cfg(test)]
mod test {
    use insta::assert_debug_snapshot;

    use crate::utils::parsing::parse_schema;

    #[test]
    fn test_strip_schema_internals() {
        use crate::consumer_schema::strip_schema_internals::StripSchemaInternals;

        let schema = r#"
            schema
  @link(url: "https://specs.apollo.dev/link/v1.0")
  @link(url: "https://specs.apollo.dev/join/v0.3", for: EXECUTION)
  @link(url: "https://specs.apollo.dev/tag/v0.3")
  @link(url: "https://specs.apollo.dev/inaccessible/v0.2", for: SECURITY)
  @link(
    url: "https://myspecs.dev/myDirective/v1.0"
    import: ["@myDirective", { name: "@anotherDirective", as: "@hello" }]
  ) {
  query: Query
}

directive @join__enumValue(graph: join__Graph!) repeatable on ENUM_VALUE

directive @join__field(
  graph: join__Graph
  requires: join__FieldSet
  provides: join__FieldSet
  type: String
  external: Boolean
  override: String
  usedOverridden: Boolean
) repeatable on FIELD_DEFINITION | INPUT_FIELD_DEFINITION

directive @join__graph(name: String!, url: String!) on ENUM_VALUE

directive @join__implements(
  graph: join__Graph!
  interface: String!
) repeatable on OBJECT | INTERFACE

directive @join__type(
  graph: join__Graph!
  key: join__FieldSet
  extension: Boolean! = false
  resolvable: Boolean! = true
  isInterfaceObject: Boolean! = false
) repeatable on OBJECT | INTERFACE | UNION | ENUM | INPUT_OBJECT | SCALAR

directive @join__unionMember(
  graph: join__Graph!
  member: String!
) repeatable on UNION

scalar join__FieldSet

directive @link(
  url: String
  as: String
  for: link__Purpose
  import: [link__Import]
) repeatable on SCHEMA

scalar link__Import

enum link__Purpose {
  """
  `SECURITY` features provide metadata necessary to securely resolve fields.
  """
  SECURITY

  """
  `EXECUTION` features provide metadata necessary for operation execution.
  """
  EXECUTION
}

directive @tag(
  name: String!
) repeatable on FIELD_DEFINITION | OBJECT | INTERFACE | UNION | ARGUMENT_DEFINITION | SCALAR | ENUM | ENUM_VALUE | INPUT_OBJECT | INPUT_FIELD_DEFINITION | SCHEMA

directive @inaccessible on FIELD_DEFINITION | OBJECT | INTERFACE | UNION | ENUM | ENUM_VALUE | SCALAR | INPUT_OBJECT | INPUT_FIELD_DEFINITION | ARGUMENT_DEFINITION

enum join__Graph {
  INVENTORY @join__graph(name: "inventory", url: "")
  PANDAS @join__graph(name: "pandas", url: "")
  PRODUCTS @join__graph(name: "products", url: "")
  REVIEWS @join__graph(name: "reviews", url: "")
  USERS @join__graph(name: "users", url: "")
}

directive @myDirective(a: String!) on FIELD_DEFINITION

directive @hello on FIELD_DEFINITION

type Product implements ProductItf & SkuItf
  @join__type(graph: INVENTORY, key: "id")
  @join__type(graph: PRODUCTS, key: "id")
  @join__type(graph: PRODUCTS, key: "sku package")
  @join__type(graph: PRODUCTS, key: "sku variation { id }")
  @join__type(graph: REVIEWS, key: "id")
  @join__implements(graph: INVENTORY, interface: "ProductItf")
  @join__implements(graph: PRODUCTS, interface: "ProductItf")
  @join__implements(graph: PRODUCTS, interface: "SkuItf")
  @join__implements(graph: REVIEWS, interface: "ProductItf") {
  id: ID! @tag(name: "hi-from-products")
  dimensions: ProductDimension
    @join__field(graph: INVENTORY, external: true)
    @join__field(graph: PRODUCTS)
  delivery(zip: String): DeliveryEstimates
    @join__field(
      graph: INVENTORY
      requires: "dimensions{...on ProductDimension{size weight}}"
    )
  sku: String @join__field(graph: PRODUCTS)
  package: String @join__field(graph: PRODUCTS)
  variation: ProductVariation @join__field(graph: PRODUCTS)
  name: String @hello @join__field(graph: PRODUCTS)
  createdBy: User @join__field(graph: PRODUCTS)
  hidden: String @join__field(graph: PRODUCTS)
  reviewsScore: Float! @join__field(graph: REVIEWS, override: "products")
  oldField: String @join__field(graph: PRODUCTS)
  reviewsCount: Int! @join__field(graph: REVIEWS)
  reviews: [Review!]! @join__field(graph: REVIEWS)
}

type ProductDimension
  @join__type(graph: INVENTORY)
  @join__type(graph: PRODUCTS) {
  size: String
  weight: Float
}

type DeliveryEstimates @join__type(graph: INVENTORY) {
  estimatedDelivery: String
  fastestDelivery: String
}

type Query
  @join__type(graph: INVENTORY)
  @join__type(graph: PANDAS)
  @join__type(graph: PRODUCTS)
  @join__type(graph: REVIEWS)
  @join__type(graph: USERS) {
  allPandas: [Panda] @join__field(graph: PANDAS)
  panda(name: ID!): Panda @join__field(graph: PANDAS)
  allProducts: [ProductItf] @join__field(graph: PRODUCTS)
  product(id: ID!): ProductItf @join__field(graph: PRODUCTS)
  review(id: Int!): Review @join__field(graph: REVIEWS)
}

type Mutation
  @join__type(graph: REVIEWS)
{
  newRandomReview: Review
}

type Panda @join__type(graph: PANDAS) {
  name: ID!
  favoriteFood: String @tag(name: "nom-nom-nom")
}

type ProductVariation @join__type(graph: PRODUCTS) {
  id: ID!
  name: String
}

type User
  @join__type(graph: PRODUCTS, key: "email")
  @join__type(graph: USERS, key: "email") {
  email: ID! @tag(name: "test-from-users")
  totalProductsCreated: Int
  name: String @join__field(graph: USERS)
}

type Review @join__type(graph: REVIEWS) {
  id: Int!
  body: String!
}

interface ProductItf implements SkuItf
  @join__type(graph: INVENTORY)
  @join__type(graph: PRODUCTS)
  @join__type(graph: REVIEWS)
  @join__implements(graph: PRODUCTS, interface: "SkuItf") {
  id: ID!
  dimensions: ProductDimension
    @join__field(graph: INVENTORY)
    @join__field(graph: PRODUCTS)
  delivery(zip: String): DeliveryEstimates @join__field(graph: INVENTORY)
  sku: String @join__field(graph: PRODUCTS)
  name: String @join__field(graph: PRODUCTS)
  package: String @join__field(graph: PRODUCTS)
  variation: ProductVariation @join__field(graph: PRODUCTS)
  createdBy: User @join__field(graph: PRODUCTS)
  hidden: String @join__field(graph: PRODUCTS) @inaccessible
  oldField: String
    @join__field(graph: PRODUCTS)
    @deprecated(reason: "refactored out")
  reviewsCount: Int! @join__field(graph: REVIEWS)
  reviewsScore: Float! @join__field(graph: REVIEWS)
  reviews: [Review!]! @join__field(graph: REVIEWS)
}

interface SkuItf @join__type(graph: PRODUCTS) {
  sku: String
}

enum ShippingClass @join__type(graph: INVENTORY) @join__type(graph: PRODUCTS) {
  STANDARD @join__enumValue(graph: INVENTORY) @join__enumValue(graph: PRODUCTS)
  EXPRESS @join__enumValue(graph: INVENTORY) @join__enumValue(graph: PRODUCTS)
  OVERNIGHT @join__enumValue(graph: INVENTORY)
}
  "#;

        let schema = graphql_parser::parse_schema(schema).unwrap();
        let schema = StripSchemaInternals::strip_schema_internals(&schema);
        let schema_str = format!("{}", schema);

        assert_debug_snapshot!(schema_str);

        parse_schema(&schema_str);
    }
}
