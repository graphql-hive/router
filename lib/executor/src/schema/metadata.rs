use std::collections::{HashMap, HashSet};

use graphql_parser::{
    query::Type,
    schema::{Definition, TypeDefinition},
};
use query_planner::consumer_schema::ConsumerSchema;

#[derive(Debug)]
pub struct SchemaMetadata<'a> {
    pub possible_types: PossibleTypes<'a>,
    pub type_fields: TypeFieldsMap<'a>,
}

type TypeFieldsMap<'a> = HashMap<&'a str, HashMap<&'a str, &'a str>>;

impl<'a> SchemaMetadata<'a> {
    pub fn new<'b: 'static>(schema: &'b ConsumerSchema) -> Self {
        schema_metadata(schema)
    }
}

#[derive(Debug, Default)]
pub struct PossibleTypes<'a> {
    map: HashMap<&'a str, HashSet<&'a str>>,
}

impl<'a> PossibleTypes<'a> {
    pub fn entity_satisfies_type_condition(
        &'a self,
        type_name: &'a str,
        type_condition: &'a str,
    ) -> bool {
        if type_name == type_condition {
            true
        } else if let Some(possible_types_of_type) = self.map.get(type_condition) {
            possible_types_of_type.contains(type_name)
        } else {
            false
        }
    }
    pub fn get_possible_types(&'a self, type_name: &'a str) -> impl Iterator<Item = &'a str> + 'a {
        self.map
            .get(type_name)
            .into_iter()
            .flatten()
            .copied()
            .chain(std::iter::once(type_name))
    }

    pub fn get_possible_types_sorted(&'a self, type_name: &'a str) -> Vec<&'a str> {
        let mut list: Vec<&'a str> = self
            .map
            .get(type_name)
            .into_iter()
            .flatten()
            .copied()
            .chain(std::iter::once(type_name))
            .collect();

        list.sort_unstable();

        list
    }
}

fn schema_metadata<'a: 'static>(schema: &'a ConsumerSchema) -> SchemaMetadata<'a> {
    let mut first_possible_types: HashMap<&str, Vec<&str>> = HashMap::new();
    let mut type_fields: HashMap<&str, HashMap<&str, &str>> = HashMap::new();

    for definition in &schema.document.definitions {
        match definition {
            Definition::TypeDefinition(TypeDefinition::Object(object_type)) => {
                let name = object_type.name.as_str();
                let fields = type_fields.entry(name).or_default();
                for field in &object_type.fields {
                    let field_type_name = field.field_type.type_name();
                    fields.insert(field.name.as_str(), field_type_name);
                }

                for interface in &object_type.implements_interfaces {
                    let interface_name = interface.as_str();
                    let possible_types_entry =
                        first_possible_types.entry(interface_name).or_default();
                    possible_types_entry.push(object_type.name.as_str());
                }
            }
            Definition::TypeDefinition(TypeDefinition::Interface(interface_type)) => {
                let name = interface_type.name.as_str();
                let mut fields = HashMap::new();
                for field in &interface_type.fields {
                    let field_type_name = field.field_type.type_name();
                    fields.insert(field.name.as_str(), field_type_name);
                }
                type_fields.insert(name, fields);
                for interface_name in &interface_type.implements_interfaces {
                    let interface_name = interface_name.as_str();
                    let possible_types_entry =
                        first_possible_types.entry(interface_name).or_default();
                    possible_types_entry.push(interface_type.name.as_str());
                }
            }
            Definition::TypeDefinition(TypeDefinition::Union(union_type)) => {
                let name = union_type.name.as_str();
                let mut types = vec![];
                for member in &union_type.types {
                    types.push(member.as_str());
                }
                first_possible_types.insert(name, types);
            }
            _ => {}
        }
    }

    let mut final_possible_types: HashMap<&str, HashSet<&str>> = HashMap::new();
    // Re-iterate over the possible_types
    for (definition_name_of_x, first_possible_types_of_x) in &first_possible_types {
        let mut possible_types_of_x: HashSet<&str> = HashSet::new();
        for definition_name_of_y in first_possible_types_of_x {
            possible_types_of_x.insert(definition_name_of_y);
            let possible_types_of_y = first_possible_types.get(definition_name_of_y);
            if let Some(possible_types_of_y) = possible_types_of_y {
                for definition_name_of_z in possible_types_of_y {
                    possible_types_of_x.insert(definition_name_of_z);
                }
            }
        }
        final_possible_types.insert(definition_name_of_x, possible_types_of_x);
    }

    SchemaMetadata {
        possible_types: PossibleTypes {
            map: final_possible_types,
        },
        type_fields,
    }
}

trait TypeName<'a> {
    fn type_name(&'a self) -> &'a str;
}

impl<'a> TypeName<'a> for Type<'a, String> {
    fn type_name(&'a self) -> &'a str {
        match self {
            graphql_parser::schema::Type::NamedType(named_type) => named_type.as_str(),
            graphql_parser::schema::Type::NonNullType(non_null_type) => non_null_type.type_name(),
            graphql_parser::schema::Type::ListType(list_type) => list_type.type_name(),
        }
    }
}
