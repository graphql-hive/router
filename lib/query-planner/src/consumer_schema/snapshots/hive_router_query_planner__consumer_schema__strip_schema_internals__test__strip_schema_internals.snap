---
source: lib/query-planner/src/consumer_schema/strip_schema_internals.rs
expression: schema_str
---
"schema {\n  query: Query\n}\n\ndirective @myDirective(a: String!) on FIELD_DEFINITION\n\ndirective @hello on FIELD_DEFINITION\n\ntype Product implements ProductItf & SkuItf {\n  id: ID!\n  dimensions: ProductDimension\n  delivery(zip: String): DeliveryEstimates\n  sku: String\n  package: String\n  variation: ProductVariation\n  name: String @hello\n  createdBy: User\n  hidden: String\n  reviewsScore: Float!\n  oldField: String\n  reviewsCount: Int!\n  reviews: [Review!]!\n}\n\ntype ProductDimension {\n  size: String\n  weight: Float\n}\n\ntype DeliveryEstimates {\n  estimatedDelivery: String\n  fastestDelivery: String\n}\n\ntype Query {\n  allPandas: [Panda]\n  panda(name: ID!): Panda\n  allProducts: [ProductItf]\n  product(id: ID!): ProductItf\n  review(id: Int!): Review\n}\n\ntype Mutation {\n  newRandomReview: Review\n}\n\ntype Panda {\n  name: ID!\n  favoriteFood: String\n}\n\ntype ProductVariation {\n  id: ID!\n  name: String\n}\n\ntype User {\n  email: ID!\n  totalProductsCreated: Int\n  name: String\n}\n\ntype Review {\n  id: Int!\n  body: String!\n}\n\ninterface ProductItf implements SkuItf {\n  id: ID!\n  dimensions: ProductDimension\n  delivery(zip: String): DeliveryEstimates\n  sku: String\n  name: String\n  package: String\n  variation: ProductVariation\n  createdBy: User\n  hidden: String\n  oldField: String @deprecated(reason: \"refactored out\")\n  reviewsCount: Int!\n  reviewsScore: Float!\n  reviews: [Review!]!\n}\n\ninterface SkuItf {\n  sku: String\n}\n\nenum ShippingClass {\n  STANDARD\n  EXPRESS\n  OVERNIGHT\n}\n"
